From 0bc5a15a613ad4e25062da60b021673f28df5b4b Mon Sep 17 00:00:00 2001
From: Johan Hovold <jhovold@gmail.com>
Date: Thu, 2 Jan 2014 22:49:22 +0100
Subject: [PATCH 3/5] USB: pl2303: only wake up MSR queue on changes
Git-Commit: dbfd2866ac36bab52a4f37384e935eb5df76ad60
Patch-Mainline: v3.14
References: bnc#959649

Only wake up MSR wait queue on actual modem-status changes.

Signed-off-by: Johan Hovold <jhovold@gmail.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Conflicts:
	drivers/usb/serial/pl2303.c
Signed-off-by: Oliver Neukum <oneukum@suse.com>
---
 drivers/usb/serial/pl2303.c | 16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

diff --git a/drivers/usb/serial/pl2303.c b/drivers/usb/serial/pl2303.c
index 97c6fe5..794eccb 100644
--- a/drivers/usb/serial/pl2303.c
+++ b/drivers/usb/serial/pl2303.c
@@ -131,6 +131,7 @@ static struct usb_driver pl2303_driver = {
 #define VENDOR_READ_REQUEST		0x01
 
 #define UART_STATE			0x08
+#define UART_STATE_MSR_MASK		0x8b
 #define UART_STATE_TRANSIENT_MASK	0x74
 #define UART_DCD			0x01
 #define UART_DSR			0x02
@@ -728,15 +729,18 @@ static void pl2303_update_line_status(struct usb_serial_port *port,
 	spin_unlock_irqrestore(&priv->lock, flags);
 	if (priv->line_status & UART_BREAK_ERROR)
 		usb_serial_handle_break(port);
-	wake_up_interruptible(&priv->delta_msr_wait);
 
-	if (delta & UART_DCD) {
-		tty = tty_port_tty_get(&port->port);
-		if (tty) {
-			usb_serial_handle_dcd_change(port, tty,
+	if (delta & UART_STATE_MSR_MASK) {
+		if (delta & UART_DCD) {
+			tty = tty_port_tty_get(&port->port);
+			if (tty) {
+				usb_serial_handle_dcd_change(port, tty,
 							status & UART_DCD);
-			tty_kref_put(tty);
+				tty_kref_put(tty);
+			}
 		}
+
+		wake_up_interruptible(&port->port.delta_msr_wait);
 	}
 }
 
-- 
2.1.4

